template pull request

myPar opened the pull request.

Commit:
commit tasks to new branch "solution"

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+str = ""
```
Надо оформить это отдельной функцией, принимающей число до которого считать (100), возвращающей строку (в твоем случае).

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+
+for i in range(100):
+    j = i + 1
+    if (j % 3 == 0) and (j % 5 == 0):
```
Скобки не нужны.

ddyak replied:
Если число делится и на 3 и на 5, то оно делится на 15. Можно сократить в целом.

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+str = ""
+
+for i in range(100):
+    j = i + 1
```
Зачем заводить j, можно работать с i? Я так понимаю беспокоит отставание на единицу? Цикл for можно сделать по полуинтервалу [a,b)

```
for i in range(a, b):
   pass
```

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+            if j % 5 == 0:
+                str += "Buzz"
+            else:
+                str += j.__str__()
```
Функции с двумя подчерквиваниями лучше не использовать, поскольку они являются своего рода приватными. Надо ```str(j)```.

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+                str += j.__str__()
+    str += " "
+
+print(str)
```
Сделать main с вводом числа и вызовом функции.

ddyak commented on FizzBuzz.py:
```diff
@@ -0,0 +1,17 @@
+    j = i + 1
+    if (j % 3 == 0) and (j % 5 == 0):
+        str += "FizzBuzz"
+    else:
```
Лучше переписать через: if-elif-elif-else

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+import math
+
+def average(list):
```
-> float

ddyak replied:
Не указал типы. Для list тут тип List[float]. Почитать можно тут: https://docs.python.org/3/library/typing.html#type-aliases

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+        number += (list[i] - aver)**2
+    return math.sqrt(number/(len(list) - 1))
+
+def get_median(list):
```
Это не медианное значение, а значение в середине массива. 
https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D0%B4%D0%B8%D0%B0%D0%BD%D0%B0_(%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%81%D1%82%D0%B8%D0%BA%D0%B0)

Иначе говоря. Забыл отсортировать.

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+def solution(list):
+    if len(list) == 0:
+        return "the list is empty"
+    min_element = min(list)
```
Это верно. Но в учебных целях нужно реализовать функцию min/max самому. (Сделай какую-нибудь одну).

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+            number = 1
+        prev_numb = list[i]
+    if max_number <= 1:
+        return 0
```
Для [1, 2, 3] вернет 0. Тогда как поряд идущих одинаковых элементов 1.

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+    for i in range(len(list)):
+        if list[i] == prev_numb:
+            number += 1
+        if list[i] != prev_numb or i >= len(list) - 1:
```
Переписать через if-else? Сложно читать.

myPar replied:
Если мы введём список, на конце которого будут одинаковые элементы (например [2, 5, ..., 3, 3, 3]), то когда i будет равно последнему индексу списка, второе условие, необходимое для обновления max_number, не будет выполнено при конструкции else if. Оба условия должны выполняться независимо.

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+
+def solution(list):
+    if len(list) == 0:
+        return "the list is empty"
```
Пока не проходили, но лучше кидать exception в таких целях. 
```
raise "The list is empty"
```

ddyak commented on ArraySpec.py:
```diff
@@ -0,0 +1,97 @@
+        standard_dev = None
+    equals = get_sameNubers(list)
+    monotone = max(non_decr(list), non_incr(list))
+    return(min_element, max_element, aver, median, standard_dev, equals, monotone)
```
Скобки для tuple лишние. Но если хочется их оставить, то надо отделять от return. Исправляй замечания от IDE.

ddyak commented on DayNumber.py:
```diff
@@ -0,0 +1,53 @@
+            return -1
+
+
+d_leap = {1: 31, 2: 29, 3: 31, 4: 30, 5: 31, 6: 30, 7: 31, 8: 31, 9: 30, 10: 31, 11: 30, 12: 31}  # leap year day dict
```
Словари принято оформлять в столбик:
```
d_leap = {1: 31,
         2: 29,
         .....}
```

ddyak replied:
Если заводишь глобальные константы, то их нужно объявлять выше функции, в которых их используешь. У тебя только в одном месте они используются, следовательно, можно поместить эти словари в функцию, чтобы не портить глобальную область видимости.

ddyak commented on DayNumber.py:
```diff
@@ -0,0 +1,53 @@
+def correct_input(day, month, year, day_dict):  # function checks for incorrect input
+    if day < 1 or (month < 1 or month > 12) or year < 0:
+        return False
+    if day > day_dict.get(month):
```
Почему тут через .get(), а выше по оператору []?

ddyak commented on DayNumber.py:
```diff
@@ -0,0 +1,53 @@
+    return True
+
+
+def solution(day_number, m, y):  # function for call in tests
```
1. Это не solution, а check_date. Не давай общих имен функциям. Общие имена хуже чем длинные
2. Либо d,m,y, либо day, month, year. Нужно соблюдать один стиль.

ddyak commented on Nod.py:
```diff
@@ -0,0 +1,21 @@
+def nod(a: int, b: int) -> int:
```
Падает nod(-2, 1).

ddyak replied:
Не надо давать имя транслитом. Функция должна называться gcd.

ddyak commented on tests/tests.py:
```diff
@@ -0,0 +1,41 @@
+# Day number tests:
```
В целом идея неплохая. На будущее я бы рекомендовал тесты писать к каждой задаче независимо. Общий файл имел бы смысл, если бы ты тестировал модуль или библиотеку, но тут у тебя все задачи разные и приходится прыгать между файлами, чтобы не терять контекст.

 Ну и для того чтобы запустить тесты для одной задачи, придется запустить для всех задач. Если что-то упадет, то некрасиво получится, а допустим, к твоей задаче отношения не имеет.

ddyak commented on SimpleSort.py:
```diff
@@ -0,0 +1,29 @@
+def sort(list):
```
bubble_sort

ddyak commented on SimpleSort.py:
```diff
@@ -0,0 +1,29 @@
+def sort(list):
+    b = True
+
+    while(b == True):
```
1. Лишние скобки.  
2. Надо писать b is True, а не b == True

ddyak commented on SimpleSort.py:
```diff
@@ -0,0 +1,29 @@
+def sort(list):
+    b = True
```
Идея с флагом хорошая. Но имя для него плохое.

ddyak commented on SimpleSort.py:
```diff
@@ -0,0 +1,29 @@
+    while(b == True):
+        b = False
+
+        for i in range(len(list) - 1):
```
Почему -1?

myPar replied:
Потому что сравниваются элементы list[i] и list[i+1].  Без -1 будет выход за границы массива.

ddyak commented on SimpleSort.py:
```diff
@@ -0,0 +1,29 @@
+                b = True
+
+                number = list[i]
+                list[i] = list[i + 1]
```
Питон позволяет писать 

```
list[i], list[i+1] = list[i+1], list[i]
```

Commit:
commit after review
...

Commit:
added tasks: "Fibonacci", "All permutations", "Quick sort".

"Quick sort" tests include generation of input lists with random float elements; lists length coming with input parameter.

Commit:
added tasks: "Heap sort", "Merge sort"

Commit:
added tasks: "Choose sort" and "Paste sort"
added tests for all sorts

Commit:
added time benchmarks. Wrote function which gets time-lists of sorting equal arrays from size 1 to "max_arr_len" for quadratic sorts. Did not plot anything on graphics yet.

Commit:
added build time diagram functions.
Plots time graphic of sorting arrays with different length by different sorts (plots all types of benchmarks)

ddyak commented on Fibonacci.py:
```diff
@@ -0,0 +1,31 @@
+    return fib_rec(number - 1) + fib_rec(number - 2)
+
+
+def fib_dynamic(number: int) -> int:
```
Еще есть интересное решение через рекурсию, но с запоминанием. https://www.python-course.eu/python3_memoization.php

ddyak commented on sorts/tests/test_quicksort.py:
```diff
@@ -0,0 +1,34 @@
+
+
+# tests for quick sort solution
+assert (quick_sort.solution([2, 3, 1, 44, 5, 2, 6, 64, 34, 11]) == [1, 2, 2, 3, 5, 6, 11, 34, 44, 64])
```
Хардкодить упорядоченный массив это плохо. Почему бы один и тот же массив не передавать своей сортировке и сравнивать его же с питоновским sorted()? У тебя так в merge_sort сделано.

ddyak commented on sorts/tests/test_choose_sort.py:
```diff
@@ -0,0 +1,25 @@
+import copy
+from sorts import choose_sort
```
Точно не choose. Лучше смотреть в википедии общепринятые названия. Потом когда-нибудь (на том же экзамене) спросят как реализовать selection sort, и просто не поймешь что от тебя хотят, хотя решение знаешь.

ddyak commented on sorts/tests/list_generator.py:
```diff
@@ -0,0 +1,10 @@
+def get_random_list(length: int) -> List[float]:  # generates random list with fixed length
+    random_list = []
+
+    for i in range(length):
```
Нормальное решение. Но в будущем лучше пользоваться бибиотечными функциями.

ddyak replied:
Советую изучить Numpy. В связки с matplotlib очень удобная. Это основные библиотеки в Python, без них невозможен любой серьезный data science, физический анализ экспериментов и т.д.

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+            time_dict["heap_sort"].append(sorted_benchmark(heap_sort.heap_sort, copy3))
+
+    x_list = np.arange(1, max_arr_len + 1, 1)
+    y_list1 = time_dict["quick_sort"]
```
Вместо (x, y_list_i) лучше использовать настоящие имена, что-то из (size_list, time_sort_i)

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+            time_dict["merge_sort"].append(sorted_benchmark(merge_sort.merge_sort, copy2))
+            time_dict["heap_sort"].append(sorted_benchmark(heap_sort.heap_sort, copy3))
+
+    x_list = np.arange(1, max_arr_len + 1, 1)
```
Numpy. Круто.  

Лучше np.arange(1, max_arr_len). Параметры по умолчанию не стоит писать.

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+    for i in range(1, max_arr_len + 1):  # run benchmarks on arrays with different length, from 1 to max_arr_len
+        input_arr = get_random_list(i)  # get list of length i with random elements
+
+        copy1 = copy.copy(input_arr)
```
Нехорошее дублирование кода. Подумай как это красиво решить. 

Мне видится удобным, чтобы входной массив не изменялся, а на выходе получался новый сортированный массив. Тогда все эти копирования (если потребуются), уйдут в функции сортировок, а пользователю кода не надо будет каждый раз копировать свой массив.

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+    return x_list, [y_list1, y_list2, y_list3]
+
+
+def plot_graphics(x_list: object, y_lists: List[List[float]], colors, labels: List[object], axis):
```
Одноразовая фунция из двух строчек? Так ли она нужна. Обычно если код с графикой matplotlib несложный (а у тебя он несложный), то пишится он цельным блоком в одной функции. У тебя это plot_builds.

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+    ax1.set_ylabel('time s')
+    ax1.grid()
+    ax1.set_title('on arbitrary arrays')
+    print(ax1.title)
```
Если print не нужен, то его нужно удалить

ddyak commented on sorts/benchmarks/time_benchmark.py:
```diff
@@ -0,0 +1,15 @@
+    return abs(end_time - start_time)    # return delta time in micro seconds
+
+
+def sorted_benchmark(sort_function, input_list: List[float]):  # benchmark for reversed sorted arrays
```
Разве нельзя просто sorted() передать в функцию выше?

benchmark(sorted, input_list).

ddyak replied:
sort() это только метод массива. 

sorted() глобальная функция, принимающая что угодно (что можно сортировать).

ddyak commented on sorts/choose_sort.py:
```diff
@@ -0,0 +1,25 @@
+def choose_sort(input_list: List[float]) -> List[float]:
+    for i in range(len(input_list)):
+        min_element = input_list[i]
+        min_ind = 0
```
ind -> idx (так общепринято называть индексы)

ddyak commented on sorts/choose_sort.py:
```diff
@@ -0,0 +1,25 @@
+        min_element = input_list[i]
+        min_ind = 0
+
+        for j in range(i, len(input_list)):
```
Можно упростить весь цикл. Все-таки как найти min ты уже показывал.

```
slice = input_list[i:]
min_element = np.min(slice)
min_idx = np.argmin(slice)
```

ddyak commented on sorts/merge_sort.py:
```diff
@@ -0,0 +1,39 @@
+        else:
+            output_list.append(list2[ind2])
+            ind2 += 1
+    if ind1 < len(list1):
```
Проверки правильные, на С/C++ надо будет явно так писать. Но в питоне можно сделать так вместо 4х строк.

```
output_list += list[ind1:] + list[ind2:]
```

Прибавление пустого массива не изменяет массив

ddyak commented on sorts/merge_sort.py:
```diff
@@ -0,0 +1,39 @@
+    if len(input_list) == 1:
+        return input_list           # list with sie 1 is sorted yet
+    else:
+        mid_ind = int((len(input_list) - 1) / 2)
```
/ -> //

ddyak commented on sorts/merge_sort.py:
```diff
@@ -0,0 +1,39 @@
+    else:
+        mid_ind = int((len(input_list) - 1) / 2)
+        left_list = merge_sort(input_list[0:mid_ind + 1])                   # get left sorted list
+        right_list = merge_sort(input_list[mid_ind + 1:len(input_list)])    # get right sorted list
```
Дефолтные значения не надо указывать. Тут это нули и размеры массива.

```
left_list = merge_sort(input_list[:mid_ind + 1])                   # get left sorted list
right_list = merge_sort(input_list[mid_ind + 1:])
```

ddyak commented on sorts/paste_sort.py:
```diff
@@ -0,0 +1,26 @@
+                    input_list[j], input_list[j - 1] = input_list[j - 1], input_list[j]  # swap two elements
+                else:
+                    break
+                j -= 1
```
Лучше занести в if блок, выразительнее. После else мы все равно сюда не попадем из-за break.

ddyak commented on sorts/quick_sort.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def quick_sort(left_ind: int, right_ind: int, input_list: List[float]):
+    if left_ind < right_ind:  # while left index less then right index do recursion
```
Много комментариев не по делу. И так видно как работает код, не надо дублировать его в комментариях. Обычно для подобных разъяснений используется документация (комментарий вначале функции с 3 кавычками)

ddyak commented on sorts/benchmarks/sort_speed_build.py:
```diff
@@ -0,0 +1,110 @@
+def build_NlogN_sorts(max_arr_len: int, is_sorted: bool):
+    time_dict = {"quick_sort": [], "merge_sort": [], "heap_sort": []}
+
+    for i in range(1, max_arr_len + 1):  # run benchmarks on arrays with different length, from 1 to max_arr_len
```
Нужны больше размеры массивов. (и шаг побольше, поскольку будешь долго считать)

Commit:
fixed remarks on last commit.
(added time benchmarking of standard sort)
